src/server/session.ts

```diff
@@ -304,6 +286,106 @@ namespace ts.server {
+                    outputs.push(output);
+                }
+            }
+        });
+        return outputs;
+    }
+
+    function combineProjectOutputForRenameLocations(projects: Projects, initialLocation: sourcemaps.SourceMappableLocation, projectService: ProjectService, findInStrings: boolean, findInComments: boolean): ReadonlyArray<RenameLocation> {
+        const outputs: RenameLocation[] = [];
+
+        combineProjectOutputWorker(projects, initialLocation, projectService, ({ project, location }, tryAddToTodo) => {
+            for (const output of project.getLanguageService().findRenameLocations(location.fileName, location.position, findInStrings, findInComments) || emptyArray) {
+                if (!contains(outputs, output, documentSpansEqual) && !tryAddToTodo(project, documentSpanLocation(output))) {
+                    outputs.push(output);
+                }
+            }
+        });
+
+        return outputs;
+    }
+
+    function combineProjectOutputForReferences(projects: Projects, initialLocation: sourcemaps.SourceMappableLocation, projectService: ProjectService): ReadonlyArray<ReferencedSymbol> {
+        const outputs: ReferencedSymbol[] = [];
+
+        combineProjectOutputWorker(projects, initialLocation, projectService, ({ project, location }, tryAddToTodo) => {
+            for (const outputReferencedSymbol of project.getLanguageService().findReferences(location.fileName, location.position) || emptyArray) {
+                let symbolToAddTo = find(outputs, o => documentSpansEqual(o.definition, outputReferencedSymbol.definition));
+                if (!symbolToAddTo) {
+                    symbolToAddTo = { definition: outputReferencedSymbol.definition, references: [] };
+                    outputs.push(symbolToAddTo);
+                }
+
+                for (const ref of outputReferencedSymbol.references) {
+                    if (!contains(symbolToAddTo.references, ref, documentSpansEqual) && !tryAddToTodo(project, documentSpanLocation(ref))) {
+                        symbolToAddTo.references.push(ref);
+                    }
+                }
+            }
+        });
+
+        return outputs.filter(o => o.references.length !== 0);
+    }
+
+    interface ProjectAndLocation {
+        readonly project: Project;
+        readonly location: sourcemaps.SourceMappableLocation;
+    }
+
+    function combineProjectOutputWorker(
+        projects: Projects,
+        initialLocation: sourcemaps.SourceMappableLocation,
+        projectService: ProjectService,
+        cb: (where: ProjectAndLocation, getMappedLocation: (project: Project, location: sourcemaps.SourceMappableLocation) => boolean) => void,
+    ): void {
+        const seenProjects = createMap<true>();
+        const toDo: ProjectAndLocation[] = (isProjectsArray(projects) ? projects : projects.projects).map(project => ({ project, location: initialLocation }));

```

Comment:
I would avoid creating this array if there is no sourceMaps to navigate to.. That would save current scenarios to create these arrays unnecessarily and would need arrays only if there are sourceMap. (original project and symLinkedProjects are unique so you can directly add then to seenMap.

Reply:
That doesn't seem worth it to me... an array with a single element is probably smaller than an `Identifier`.

Reply:
If you pull out calling of cb to a function you wont need to have closures and will not need the creation of new array if not needed.

eg.
```
 function callbackProjectAndLocation(
        projectAndLocation: ProjectAndLocation,
        projectService: ProjectService,
        seenProjects: Map<true>,
        toDo: ProjectAndLocation[] | undefined,
        cb: (where: ProjectAndLocation, getMappedLocation: (project: Project, location: sourcemaps.SourceMappableLocation) => boolean) => void,
    ) : ProjectAndLocation[] | undefined 
```

Reply:
In fact you could delay creation of seenProjects as well since that is also needed only when sourceMapped location is found.